### PR TITLE
Misc fixes working with plans

### PR DESCRIPTION
**Pull request by ahgittin**
Change CAMP parsing so it can load items which use other transformers.  Previously it assumed referenced types were either java or yaml; now it uses the plan transformer extensions.  Messier than I had thought so please review commits.

Misc other tidies including errors and generics, as described in commits.

### CODE REVIEW

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/BasicBrooklynCatalog.java [hunk 1]
```diff
@@ -800,7 +809,11 @@ private boolean attemptType(String key, CatalogItemType candidateCiType) {
                     // (when we're not given a key, the previous block should apply)
                     errors.add(e);
                 } else {
-                    // all other cases, the error is probably due to us not getting the type right, ignore it
+                    // all other cases, the error is probably due to us not getting the type right, probably ignore it
+                    // but cache it if we've checked entity, we'll use that as fallback errors
```
Comment (aledsage) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/BasicBrooklynCatalog.java]: Not sure I follow - why would we get the type wrong for non-entities? Have we guessed that it was an entity, without being able to tell from the context if its instead a policy/enricher/etc?

File: core/src/main/java/org/apache/brooklyn/core/plan/PlanToSpecFactory.java [hunk 2]
```diff
@@ -95,13 +96,18 @@ public static PlanToSpecTransformer forPlanType(ManagementContext mgmt, String p
...
+                if (result==null) {
+                    transformersWhoDontSupport.add(t.getShortDescription() + " (returned null)");
+                    continue;
+                }
+                return Maybe.of(result);
             } catch (PlanNotRecognizedException e) {
                 transformersWhoDontSupport.add(t.getShortDescription() +
                     (Strings.isNonBlank(e.getMessage()) ? " ("+e.getMessage()+")" : ""));
-            } catch (Exception e) {
+            } catch (Throwable e) {
```
Comment (aledsage) [inline core/src/main/java/org/apache/brooklyn/core/plan/PlanToSpecFactory.java]: Out of interest, why catch `Throwable`? We immediately do a `Exceptions.propagateIfFatal(e)` so if it was an `Error` then we'd rethrow immediately (as though we hadn't caught). Are there other throwable types that we're hitting?

(I'm fine with the change; just want to know what throwables we should expect to encounter).

File: utils/common/src/test/java/org/apache/brooklyn/util/time/TimeTest.java [hunk 3]
```diff
@@ -152,7 +152,13 @@ public void testElapsedSince() {
         Assert.assertFalse(Time.hasElapsedSince(aFewSecondsAgo, Duration.TEN_SECONDS));
         Assert.assertTrue(Time.hasElapsedSince(-1, Duration.TEN_SECONDS));
     }
-    
+
+    @Test
+    public void testParseTime() {
+        Assert.assertEquals(Time.parseElapsedTime("1s"), 1000);
```
Comment (aledsage) [inline utils/common/src/test/java/org/apache/brooklyn/util/time/TimeTest.java]: There's a lot of logic in `parseElapsedTimeAsDouble`. Would be good to test more scenarios, e.g. the inverse of of the tests in `testMakeStringExact_secondsAndMillis()`.

**Comment by aledsage** [comment]
@ahgittin looks good - only very minor comments/questions. Happy for this to be merged.

File: utils/common/src/test/java/org/apache/brooklyn/util/time/TimeTest.java (hunk 3 above)
Comment (ahgittin) [inline utils/common/src/test/java/org/apache/brooklyn/util/time/TimeTest.java]: agree

File: core/src/main/java/org/apache/brooklyn/core/plan/PlanToSpecFactory.java (hunk 2 above)
Comment (ahgittin) [inline core/src/main/java/org/apache/brooklyn/core/plan/PlanToSpecFactory.java]: i think i was hoping to catch some weird errors, didn't realize propagateIfFatal would grab all `Error`.  maybe for some things like class linking etc we don't treat them as fatal?

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/BasicBrooklynCatalog.java (hunk 1 above)
Comment (ahgittin) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/BasicBrooklynCatalog.java]: we're parsing CAMP, then CAMP refers to something in catalog.  the code before this change would use camp-specific logic to handle the catalog entry making use of the parent entity's class loading context if needed.  but that breaks if it refers to a non-camp non-java catalog item (eg tosca).  so this now also looks it up using the catalog mechanism (although in that case it does not have any default class loading context.)

the CLC should only matter for java types.  it's working now, but would like to clean this whole area up massively!

**Pull request merged**